File: session-04-proxy-and-aop/go/order/service.go
package order

// OrderService interface: this is what the proxy wraps.
// Both the real service and the proxy implement this interface.
type OrderService interface {
	ListOrders() []Order
	FindOrder(id int) *Order
}

// orderServiceImpl: the real service. NO logging, NO timing.
type orderServiceImpl struct {
	repo OrderRepository
}

func NewOrderService(repo OrderRepository) OrderService {
	return &orderServiceImpl{repo: repo}
}

func (s *orderServiceImpl) ListOrders() []Order {
	return s.repo.FindAll()
}

func (s *orderServiceImpl) FindOrder(id int) *Order {
	return s.repo.FindByID(id)
}

File: session-04-proxy-and-aop/python/order/service.py
from order.model import Order
from order.repository import OrderRepository
from typing import Optional


# Notice: NO logging, NO timing code here. Decorators add it.
class OrderService:
    def __init__(self, repo: OrderRepository):
        self.repo = repo

    def list_orders(self) -> list[Order]:
        return self.repo.find_all()

    def find_order(self, order_id: int) -> Optional[Order]:
        return self.repo.find_by_id(order_id)
